lace/src/interface/oracle/traits.rs

```diff
@@ -2046,6 +2066,125 @@ pub trait OracleT: CanOracle {
+            .iter()
+            .map(|state| {
+                let view_ix = state.asgn.asgn[col_ix];
+                let weights =
+                    &utils::given_weights(&[state], &[col_ix], &given)[0];
+
+                // combine the state weights with the given weights
+                let mut mm_weights: Vec<f64> = state.views[view_ix]
+                    .weights
+                    .iter()
+                    .zip(weights[&view_ix].iter())
+                    .map(|(&w1, &w2)| w1 + w2)
+                    .collect();
+
+                let z: f64 = logsumexp(&mm_weights);
+                mm_weights.iter_mut().for_each(|w| *w = (*w - z).exp());
+
+                state.views[view_ix].ftrs[&col_ix].to_mixture(mm_weights)
+            })
+            .collect();
+
+        enum MType {
+            Gaussian,
+            Categorical,
+            Count,
+            Unsupported,
+        }
+
+        let mtype = match mixture_types[0] {
+            MixtureType::Gaussian(_) => MType::Gaussian,
+            MixtureType::Poisson(_) => MType::Count,
+            MixtureType::Categorical(_) => MType::Categorical,
+            _ => MType::Unsupported,
+        };
+
+        match mtype {
+            MType::Gaussian => {
+                let mms: Vec<_> = mixture_types
+                    .drain(..)
+                    .map(|mt| {
+                        if let MixtureType::Gaussian(mm) = mt {
+                            mm
+                        } else {
+                            panic!("Expected Gaussian Mixture Type")
+                        }
+                    })
+                    .collect();
+                let mm = Mixture::combine(mms);
+                Ok(Variability::Variance(mm.variance().unwrap()))
+            }
+            MType::Count => {
+                let mms: Vec<_> = mixture_types
+                    .drain(..)
+                    .map(|mt| {
+                        if let MixtureType::Poisson(mm) = mt {
+                            mm
+                        } else {
+                            panic!("Expected Poisson Mixture Type")
+                        }
+                    })
+                    .collect();
+                let mm = Mixture::combine(mms);
+                Ok(Variability::Variance(mm.variance().unwrap()))
+            }
+            MType::Categorical => {
+                let mms: Vec<_> = mixture_types
+                    .drain(..)
+                    .map(|mt| {
+                        if let MixtureType::Categorical(mm) = mt {
+                            mm
+                        } else {
+                            panic!("Expected Categorical Mixture Type")
+                        }
+                    })
+                    .collect();
+                let mm = Mixture::combine(mms);
+                Ok(Variability::Entropy(mm.entropy()))
+            }
+            _ => panic!("Unsupported MType"),
+        }
```

Comment:
I doubt it matters in the final compilation, but is there a reason for adding the inner Enum and using that as the result of the match and then matching on the Enum again, instead of just putting the code in the original `match` statement?

Reply:
Borrow checker. If I match on `MxtureTypes[0]` I can't take `mm` out of the `MixtureType` enum variant, or at least that is what my mental model of the borrow checker told me. I didn't actually code it the other way and let the borrow checker scold me.